File: javascript/socketd/src/socketd/cluster/LoadBalancer.ts
import {ClientSession} from "../transport/client/ClientSession";
import {IoSupplier} from "../transport/core/Typealias";

/**
 * 负载均衡器
 *
 * @author noear
 * @since 2.3
 */
export class LoadBalancer {
    //轮环计数器
    private static roundCounter: number = 0;

    private static roundCounterGet(): number {
        let counter = LoadBalancer.roundCounter++;
        if (counter > 999_999) {
            LoadBalancer.roundCounter == 0;
        }
        return counter;
    }

    private static hashcode(str): number {
        var hash = 0, i, chr, len;
        if (str.length === 0) return hash;
        for (i = 0, len = str.length; i < len; i++) {
            chr = str.charCodeAt(i);
            hash = ((hash << 5) - hash) + chr;
            hash |= 0; // Convert to 32bit integer
        }
        return hash;
    }

    /**
     * 根据 poll 获取任意一个
     */
    static getAnyByPoll<T extends ClientSession>(coll: Set<T> | null): T | null {
        return LoadBalancer.getAny(coll, () => LoadBalancer.roundCounterGet());
    }

    /**
     * 根据 hash 获取任意一个
     */
    static getAnyByHash<T extends ClientSession>(coll: Set<T> | null, diversion: string): T | null {
        return LoadBalancer.getAny(coll, () => LoadBalancer.hashcode(diversion));
    }


    /**
     * 获取任意一个
     */
    static getAny<T extends ClientSession>(coll: Set<T> | null, randomSupplier: IoSupplier<number>): T | null {
        if (coll == null || coll.size == 0) {
            return null;
        } else {
            //查找可用的会话
            let sessions = new Array<T>();
            for (let s of coll) {
                if (s.isValid() && !s.isClosing()) {
                    sessions.push(s);
                }
            }

            if (sessions.length == 0) {
                return null;
            }

            if (sessions.length == 1) {
                return sessions[0];
            }

            //随机获取
            let random = Math.abs(randomSupplier());
            let idx = random % sessions.length;
            return sessions[idx];
        }
    }

    /**
     * 获取第一个
     */
    public static getFirst<T extends ClientSession>(coll: Array<T>): T | null {
        if (coll == null || coll.length == 0) {
            return null;
        } else {
            //查找可用的会话
            for (let s of coll) {
                if (s.isValid() && !s.isClosing()) {
                    return s;
                }
            }

            return null;
        }
    }
}

File: python/socketd/socketd/cluster/LoadBalancer.py
from threading import RLock
from typing import List

from socketd.transport.client.ClientSession import ClientSession
from socketd.transport.utils.StrUtil import StrUtil


class LoadBalancer:
    __roundCounter: int = 0  # 轮环计数器
    __lock: RLock = RLock()  # 计数锁

    @staticmethod
    def round_counter_get() -> int:
        LoadBalancer.__lock.acquire()

        try:
            LoadBalancer.__roundCounter += 1
            if LoadBalancer.__roundCounter > 999_999:
                LoadBalancer.__roundCounter = 0
            return LoadBalancer.__roundCounter
        finally:
            LoadBalancer.__lock.release()

    # 根据 poll 获取任意一个
    @staticmethod
    def get_any_by_poll(coll: List[ClientSession]) -> ClientSession:
        return LoadBalancer.get_any(coll, LoadBalancer.round_counter_get())

    # 根据 hash 获取任意一个
    @staticmethod
    def get_any_by_hash(coll: List[ClientSession], diversion: str) -> ClientSession:
        return LoadBalancer.get_any(coll, StrUtil.hash_code(diversion))

    # 获取任意一个
    @staticmethod
    def get_any(coll: list[ClientSession], random: int) -> ClientSession | None:
        if coll is None or coll.__len__() == 0:
            return None
        else:
            sessions: List[ClientSession] = []
            for s in coll:
                if s.is_valid() and not s.is_closing():
                    sessions.append(s)

            if sessions.__len__() == 0:
                return None

            if sessions.__len__() == 1:
                return sessions[0]

            random = abs(random)
            idx = random % sessions.__len__()
            return sessions[idx]

    # 获取第一个
    @staticmethod
    def get_first(coll: List[ClientSession]):
        if coll is None or coll.__len__() == 0:
            return None
        else:
            for s in coll:
                if s.is_valid() and not s.is_closing():
                    return s

            return None
